### PR TITLE
32bit volume control

### PR DESCRIPTION
**Pull request by maks**
This modifies the standard i2s PIO program from RPI by adding a fixed amount of bits of offset from the MSB to lower the volume on picoTracker hardware that has very  high output gain on its audio output socket.

Fixes: #13

### CODE REVIEW

**Comment by democloid** [comment]
@maks This is amazing, great work!
I haven't had a ton of time to test and it wasn't immediately obvious how to change how many bits we are displacing.
These are my observations:
1. One thing I tested is, how low can volume go? That's at 10 currently and I feel I could want to go lower than that, thoughts?
2. Why were your tests at 140 volume? Can't find exactly where now, but almost sure that 100 is 100% volume, so I think we should test at that level.
3. We don't need the original pio file, that's what git is for, no?

**Comment by democloid** [comment]
To add to this, I feel my comfortable listening level right now is between 20 and 40, and wonder if then 40 should be higher. One would like to normally listen without loosing resolution, ideally.
I think that level works fine with Line out too.
With 250Ohm headphones it does become a bit quiet, so we have to strike a balance, but I imagine that's less common?
I'd try images with 1 and 2 bits more displacement to see how it feels.

**Comment by maks** [comment]
> @maks This is amazing, great work! I haven't had a ton of time to test and it wasn't immediately obvious how to change how many bits we are displacing. These are my observations:
> 
>     1. One thing I tested is, how low can volume go? That's at 10 currently and I feel I could want to go lower than that, thoughts?
> 
>     2. Why were your tests at 140 volume? Can't find exactly where now, but almost sure that 100 is 100% volume, so I think we should test at that level.
> 
>     3. We don't need the original pio file, that's what git is for, no?


Thanks for looking at this so quickly! 👍🏻 
Yes I think you are right, the PIO code is definitely not obvious, I'll add some more comments to explain as unfort it loks like you can't have named literal values in the PIO asm syntax, so theres alot of magic numbers in there that need to be labelled.

1. Yes good point, I need to look at why vol only goes down to 10, that doesnt make sense to me either.
2. Ah I thought `140` was the default value, just realised the new project default is `100` and `140` is just what "Tardline__M-_-N" has set, so I'll test with 100 from now on. But I think `100` is 50% vol as the range of the UI setting is currently from `10` to `200`, but I'll double check what the code actually does with this value.
3. Yes I normally just delete old code, it was more for having something to easily reference, but you're right, its easy enoug to fish out of git history when we need it again. 👍🏻

**Comment by maks** [comment]
> To add to this, I feel my comfortable listening level right now is between 20 and 40, and wonder if then 40 should be higher. One would like to normally listen without loosing resolution, ideally. I think that level works fine with Line out too. With 250Ohm headphones it does become a bit quiet, so we have to strike a balance, but I imagine that's less common? I'd try images with 1 and 2 bits more displacement to see how it feels.

hmm this is a good point too! 👍🏻  I think this ends up being a matter of tweaking to find compromise value that works across a decent number of headphones. I was worried it was getting too low on my pairs which have 120Ohm when in powered noice-cancel mode. So yes, some more builds with diff values is a great idea.
@democloid if you are ok for me to merge #101 then each commit I push to this PR with a diff bit ofdset will automatically create a binary build to test out :-)

**Comment by maks** [comment]
Ha! I should learn to read docs more closely! Turns out PIO asm syntazx does have a `.define` for literals 🎉  So I'll use that to label the magic numbers and reduce deplication

**Comment by democloid** [comment]
>     1. Yes good point, I need to look at why vol only goes down to 10, that doesnt make sense to me either.
> 
>     2. Ah I thought `140` was the default value, just realised the new project default is `100` and `140` is just what "Tardline__M-_-N" has set, so I'll test with 100 from now on. But I think `100` is 50% vol as the range of the UI setting is currently from `10` to `200`, but I'll double check what the code actually does with this value.
> 

Of course I cannot find the code right now, but I'm pretty sure 100 means 100% as in, if you only play one sample that is as as loud as possible, volume 100 will play it as is, while over 100 will clip, but it's meant to increase volume for very low volume samples/mixes.
Likewise, 10 is just multiplying the samples by .1, and I imagine it's the minimum as to not loose a lot of dynamic range.

**Review by democloid — APPROVED** [review]
I think I prefer 5bit displacement. 6 is too low. Also tried the line out and I think it more or less matches other instruments I have at 100% volume (maybe a bit on the lower side, but still comparable).

File: sources/Adapters/picoTracker/audio/picoTrackerAudioDriver.cpp [hunk 1]
```diff
@@ -118,7 +118,7 @@ bool picoTrackerAudioDriver::InitDriver() {
   int sample_freq = 44100;
   // This number is exactly 20000 for our 220.5MHz core freq
```
[the remark this answers is not in the document]
Comment (maks) [reply sources/Adapters/picoTracker/audio/picoTrackerAudioDriver.cpp]: thanks for spotting! fixed.

**Pull request merged**